components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/bindings/impl/CertificateBasedTokenBinder.java

```diff
@@ -0,0 +1,212 @@
+import java.security.cert.CertificateException;
+import java.security.cert.CertificateFactory;
+import java.security.cert.X509Certificate;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import static org.wso2.carbon.identity.oauth2.OAuth2Constants.TokenBinderType.CERTIFICATE_BASED_TOKEN_BINDER;
+
+/**
+ * This class provides the certificate based token binder implementation.
+ */
+public class CertificateBasedTokenBinder extends AbstractTokenBinder {
+
+    private static TokenPersistenceProcessor hashingPersistenceProcessor;
+    private static final String RETRIEVE_TOKEN_BINDING_BY_REFRESH_TOKEN =
+            "SELECT BINDING.TOKEN_BINDING_TYPE,BINDING.TOKEN_BINDING_VALUE,BINDING.TOKEN_BINDING_REF " +
+                    "FROM IDN_OAUTH2_ACCESS_TOKEN TOKEN LEFT JOIN IDN_OAUTH2_TOKEN_BINDING BINDING ON " +
+                    "TOKEN.TOKEN_ID=BINDING.TOKEN_ID WHERE TOKEN.REFRESH_TOKEN = ? " +
+                    "AND BINDING.TOKEN_BINDING_TYPE = ?";
+
+    @Override
+    public String getBindingType() {
+
+        return CERTIFICATE_BASED_TOKEN_BINDER;
+    }
+
+    @Override
+    public List<String> getSupportedGrantTypes() {
+
+        Set<String> supportedGrantTypes = OAuthServerConfiguration.getInstance().getSupportedGrantTypes().keySet();
+        return supportedGrantTypes.stream().collect(Collectors.toList());
+    }
+
+    @Override
+    public String getDisplayName() {
+
+        return "Certificate Based";
+    }
+
+    @Override
+    public String getDescription() {
+
+        return "Bind the TLS certificate to the token. Supported grant types: Code";
+    }
+
+    @Override
+    public String getOrGenerateTokenBindingValue(HttpServletRequest request) throws OAuthSystemException {
+
+        return null;
+    }
+
+    @Override
+    public String getTokenBindingValue(HttpServletRequest request) throws OAuthSystemException {
+
+        String cnfValue = generateCnfHashValue(request);
+        if (StringUtils.isNotBlank(cnfValue)) {
+            return cnfValue;
+        } else {
+            throw new OAuthSystemException("Error occurred while generating cnf hash value.");
+        }
+    }
+
+    @Override
+    public Optional<String> getTokenBindingValue(OAuth2AccessTokenReqDTO oAuth2AccessTokenReqDTO) {
+
+        return Optional.ofNullable(generateCnfHashValue(oAuth2AccessTokenReqDTO.getHttpServletRequestWrapper()));
+    }
+
+    @Override
+    public void setTokenBindingValueForResponse(HttpServletResponse response, String bindingValue) {
+
+        // Not required.
+    }
+
+    @Override
+    public void clearTokenBindingElements(HttpServletRequest request, HttpServletResponse response) {
+
+        // Not required.
+    }
+
+    @Override
+    public boolean isValidTokenBinding(Object request, String bindingReference) {
```

Comment:
Check with IS APIs

Reply:
Tested with IS Application Management APIs with both token binding validation enabled and disabled when certificate token binding is being used.